src/lib.rs

```diff
@@ -44,6 +44,17 @@ use std::{fmt, io, ops};
 
 pub use error::Error;
 
+/// A hashing engine which bytes can be serialized into. It is expected
+/// to implment the `io::Write` trait, but to never return errors under
+/// any conditions.
+pub trait HashEngine: Clone + io::Write {
+    /// Add data to the hash engine without any error return type to deal with
+    #[inline(always)]
+    fn input(&mut self, data: &[u8]) {
+        self.write(data).expect("hash returned error");
```

Comment:
Since `write()` does not guarantee that all bytes are written, we should use `write_all()`.